### ISSUE TITLE
[Req] fit-or-vertical option for function arguments

### ISSUE THREAD

**Issue by osener** [ocaml-ppx/ocamlformat#289]
First of all, thanks for this great project! I can't stress how valuable it has been for me coming from ReasonML syntax. Thanks to ocamlformat I was able to get used to OCaml syntax very quickly and now I prefer it most of the time!

One thing I'd love to have is sparser argument list for function application and definition when it doesn't fit in the margin. I started with ocamlformat pinned to `https://github.com/janestreet/ocamlformat.git\#jane` which does this by default, and would like to switch to the official version.

As you can see in the examples below, ocamlformat master just wraps the code at margin, making it subjectively harder to scan the argument list. Same applies to argument lists at function definition.

#### Jane Street's fork:
```ocaml
let foo =
  let open Gql in
  [ field
      "id"
      ~doc:"Toy ID."
      ~args:[]
      ~typ:(non_null guid)
      ~resolve:(fun _ctx x -> x.id)
  ; field
      "name"
      ~doc:"Toy name."
      ~args:[]
      ~typ:(non_null string)
      ~resolve:(fun _ctx x -> x.name)
  ; field
      "description"
      ~doc:"Toy description."
      ~args:[]
      ~typ:string
      ~resolve:(fun _ctx x -> x.description |> Util.option_of_string)
  ; field
      "type"
      ~doc:"Toy type. Possible values are: car, animal, train."
      ~args:[]
      ~typ:(non_null toy_type_enum)
      ~resolve:(fun _ctx x -> x.toy_type)
  ; field
      "createdAt"
      ~doc:"Date created."
      ~args:[]
      ~typ:(non_null Scalar.date_time)
      ~resolve:(fun _ctx x -> x.created_at)
  ]
```
#### ocamlformat:
```ocaml
let foo =
  let open Gql in
  [ field "id" ~doc:"Toy ID." ~args:[] ~typ:(non_null guid) ~resolve:
      (fun _ctx x -> x.id )
  ; field "name" ~doc:"Toy name." ~args:[] ~typ:(non_null string) ~resolve:
      (fun _ctx x -> x.name )
  ; field "description" ~doc:"Toy description." ~args:[] ~typ:string ~resolve:
      (fun _ctx x -> x.description |> Util.option_of_string )
  ; field "type" ~doc:"Toy type. Possible values are: car, animal, train."
      ~args:[] ~typ:(non_null toy_type_enum) ~resolve:(fun _ctx x -> x.toy_type
    )
  ; field "createdAt" ~doc:"Date created." ~args:[]
      ~typ:(non_null Scalar.date_time) ~resolve:(fun _ctx x -> x.created_at )
  ]
```

**Comment by hhugo** [comment]
Implemented here https://github.com/ocaml-ppx/ocamlformat/pull/283

**Comment by hhugo** [comment]
Also, since you're talking about ReasonML, do you have any opinion on #254 ? do you ever translate files from ReasonML to ocaml ?

**Comment by osener** [comment]
@hhugo Wow! Though I shouldn't be surprised since you also implemented the version I'm using now!

But I'm actually seeing no change with the snippet I posted above. I guess it's because of how it interacts with function arguments, when I change `resolve` arg to a long string it looks exactly the way I want it.

**Comment by hhugo** [comment]
Right, a break hint is missing somewhere.  Will fix it.

**Comment by osener** [comment]
Re: #254, I think it is fair to remove Reason support from ocamlformat. I had tried using `refmt` to convert Reason sources to OCaml (and going the other way) a while back and saw some loss of syntax sugar such as `let%lwt`

I just tried again this time using ocamlformat and refmt toghether, and it works really well except the annotations Refmt inserts such as `[@reason.raw_literal ...]` and `[@explicit_arity]`. If Refmt provides an option to disable those, something like the command below would be almost lossless:
```sh
refmt --parse=ml --print=re source.ml | refmt --parse=re --print=ml | ocamlformat`

**Comment by jberdine** [comment]
@osener, which versions of refmt and ocamlformat were you using? There is code in `Reason.ml` that used to handle cleaning up `explicit_arity`, but I guess it is stale now. But the high-level point is that I think that all such attributes could be handled on the ocamlformat side without needing a special mode of refmt. Preserving the position of comments really well, however, probably takes some real work.

**Comment by osener** [comment]
@jberdine I haven't tried ocamlformat's Reason support, only refmt and refmt+ocamlformat. If handling the cases you mentioned don't add much maintenance overhead I think it would be really valuable to have. Reason has a good story for converting .ml to .re, but I can imagine many people (like me) would like to go  the other direction after discovering OCaml through Reason community.

**Comment by jberdine** [comment]
implemented in #283
